compiler/rustc_mir/src/transform/const_goto.rs

```diff
@@ -0,0 +1,129 @@
+        // make it easier for further passes
+        if should_simplify {
+            simplify_cfg(body);
+            simplify_locals(body, tcx);
+        }
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for ConstGotoOptimizationFinder<'a, 'tcx> {
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        let mut bailer = || {
+            match terminator.kind {
+                TerminatorKind::Goto { target } => {
+                    // We only apply this optimization if the last statement is a const assignment
+                    let last_statement =
+                        self.body.basic_blocks()[location.block].statements.last()?;
+
+                    match &last_statement.kind {
+                        StatementKind::Assign(box (place, Rvalue::Use(op))) => {
+                            let _const = op.constant()?;
+                            // We found a constant being assigned to `place`.
+                            // Now check that the target of this Goto switches on this place.
+                            let target_bb = &self.body.basic_blocks()[target];
+                            if !target_bb.statements.is_empty() {
+                                return None;
+                            }
+
+                            let target_bb_terminator = target_bb.terminator();
+                            match &target_bb_terminator.kind {
+                                TerminatorKind::SwitchInt { discr, switch_ty, values, targets }
+                                    if discr.place() == Some(*place) =>
+                                {
+                                    // We now know that the Switch matches on the const place, and it is statementless
+                                    // Now find which value in the Switch matches the const value.
+                                    let const_value = _const.literal.try_eval_bits(
+                                        self.tcx,
+                                        self.param_env,
+                                        switch_ty,
+                                    )?;
+                                    let found_value_idx_option = values
+                                        .iter()
+                                        .enumerate()
+                                        .find(|(_, x)| const_value == **x)
+                                        .map(|(idx, _)| idx);
+
+                                    let target_to_use_in_goto =
+                                        if let Some(found_value_idx) = found_value_idx_option {
+                                            targets[found_value_idx]
+                                        } else {
+                                            // If we did not find the const value in values, it must be the otherwise case
+                                            *targets.last()?
```

Comment:
this can't possibly fail. Maybe just unconditionally index and do `found_value_idx_option.unwrap_or(targets.len() - 1)` for the index? Otherwise, please unwrap here to signal that we aren't expecting this to ever hit `None`

Reply:
After the rebase the code is a bit different after the SwitchInt refactoring, so i'll leave it be